crates/metadata-store/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+#![allow(dead_code)]
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use restate_types::Version;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+#[derive(Debug, thiserror::Error)]
+pub enum NetworkError {
+    #[error("unreachable")]
+    Unreachable,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReadError {
+    #[error(transparent)]
+    Network(NetworkError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum WriteError {
+    #[error("failed condition: {0}")]
+    FailedCondition(String),
+    #[error(transparent)]
+    Network(NetworkError),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VersionedValue {
+    pub version: Version,
+    pub value: Bytes,
+}
+
+impl VersionedValue {
+    pub fn new(version: Version, value: Bytes) -> Self {
+        Self { version, value }
+    }
+}
+
+/// Conditions for the write operations of the [`MetadataStore`].
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Condition {
+    /// Key-value pair must not exist for the write operation to succeed.
+    DoesNotExist,
+    /// Key-value pair must have the provided [`Version`] for the write operation to succeed.
+    MatchesVersion(Version),
+}
+
+/// Metadata store abstraction. The metadata store implementations need to support linearizable
+/// reads and atomic compare and swap operations.
+#[async_trait]
+pub trait MetadataStore: Send + Sync {
+    /// Gets the value and its current version for the given key. If key-value pair is not present,
+    /// then return [`None`].
+    async fn get(&self, key: &str) -> Result<Option<VersionedValue>, ReadError>;
+
+    /// Gets the current version for the given key. If key-value pair is not present, then return
+    /// [`None`].
+    async fn get_version(&self, key: &str) -> Result<Option<Version>, ReadError>;
+
+    /// Puts the given key-value pair following the provided condition. If the operation succeeds,
+    /// then it returns the new [`Version`] of the key-value pair. If the condition is not met,
+    /// then the operation returns a [`WriteError::FailedCondition`].
+    async fn put(
+        &self,
+        key: &str,
+        value: Bytes,
+        condition: Option<Condition>,
```

Comment:
Adding a third variant (None) in `Condition` eliminates the need for Option

Reply:
True, will add it.